pkg/metrics/percentile/gk_array.go

```diff
@@ -194,6 +194,92 @@ func (s GKArray) interpolatedQuantile(q float64) float64 {
 	return weightBelow*s.Entries[indexBelow].V + weightAbove*s.Entries[indexAbove].V
 }
 
+// Quantiles accepts a sorted slice of quantile values and returns the epsilon estimates of
+// elements at those quantiles.
+func (s GKArray) Quantiles(qValues []float64) []float64 {
+	quantiles := make([]float64, 0, len(qValues))
+	if s.Count == 0 {
+		for _ = range qValues {
+			quantiles = append(quantiles, math.NaN())
+		}
+		return quantiles
+	}
+
+	// This shouldn't happen but checking just in case.
+	if len(s.Incoming) > 0 {
+		s = s.compressWithIncoming(nil)
+	}
+
+	// When there are only a few values just call interpolatedQuantile
+	// for each quantile.
+	if s.Count < int64(1/EPSILON) {
+		for _, q := range qValues {
+			if q < 0 || q > 1 {
+				quantiles = append(quantiles, math.NaN())
+			} else {
+				quantiles = append(quantiles, s.interpolatedQuantile(q))
+			}
+		}
+		return quantiles
+	}
+
+	// If the qValues are not sorted, just call Quantile for each qValue
+	if !sort.Float64sAreSorted(qValues) {
+		for k, q := range qValues {
+			quantiles[k] = s.Quantile(q)
+		}
+		return quantiles
+	}
+
+	// For sorted qValues, the quantiles can be found in one pass
+	// over the Entries
+	ranks := make([]int64, 0, len(qValues))
+	invalidQ := false
+	for _, q := range qValues {
+		if q < 0 || q > 1 {
+			invalidQ = true
+		} else {
+			ranks = append(ranks, int64(q*float64(s.Count-1)))
+		}
+	}
+
+	spread := int64(EPSILON * float64(s.Count-1))
+	gSum := int64(0)
+	i := 0
+	j := 0
+	for ; i < len(s.Entries) && j < len(ranks); i++ {
+		gSum += int64(s.Entries[i].G)
+		// Loop since adjacent ranks could be the same.
+		for ; j < len(ranks) && gSum+int64(s.Entries[i].Delta)-1 > ranks[j]+spread; j++ {
+			if i == 0 {
+				quantiles = append(quantiles, s.Min)
+			} else {
+				quantiles = append(quantiles, s.Entries[i-1].V)
+			}
+		}
+	}
+	// If there're any quantile values that have not been found,
+	// return the max value.
+	for ; j < len(ranks); j++ {
+		quantiles = append(quantiles, s.Max)
+	}
+	// Invalid qValues should return NaN
+	// This shouldn't happen, since dd controls the qValues, but leaving in
+	// just in case.
+	if invalidQ {
+		for _, q := range qValues {
+			if q < 0 {
+				quantiles = append(quantiles, 0)
+				copy(quantiles[1:], quantiles[:])
```

Comment:
I'd expect quantiles returned to always be in the same order as the qValues given, I think this breaks that assumption